Fix Material 3 Scrollable `TabBar`

TahaTesser opened the pull request:
fix https://github.com/flutter/flutter/issues/117722

### Description
1. Fix the divider doesn't stretch to take all the available width in the scrollable tab bar in M3
2. Add `dividerHeight` property.
3. Update the default tab alignment for the scrollable tab bar to match the specs (this is backward compatible for M2 with the new `tabAlignment` property).

### Bug (default tab alignment)

![Screenshot 2023-05-05 at 19 04 40](https://user-images.githubusercontent.com/48603081/236509483-1d03af21-a764-4776-acef-2126560f0d51.png)


### Fix (default tab alignment)

![Screenshot 2023-05-05 at 19 04 15](https://user-images.githubusercontent.com/48603081/236509513-2426d456-c54f-42bd-9545-a14dc6ee7e69.png)


### Code sample

<details> 
<summary>code sample</summary> 

```dart
import 'package:flutter/material.dart';

/// Flutter code sample for [TabBar].

void main() => runApp(const TabBarApp());

class TabBarApp extends StatelessWidget {
  const TabBarApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        //  tabBarTheme: const TabBarTheme(tabAlignment: TabAlignment.start),
          useMaterial3: true,
      ),
      home: const TabBarExample(),
    );
  }
}

class TabBarExample extends StatefulWidget {
  const TabBarExample({super.key});

  @override
  State<TabBarExample> createState() => _TabBarExampleState();
}

class _TabBarExampleState extends State<TabBarExample> {
  bool rtl = false;

  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      initialIndex: 1,
      length: 3,
      child: Directionality(
        textDirection:  rtl ? TextDirection.rtl : TextDirection.ltr,
        child: Scaffold(
          appBar: AppBar(
            title: const Text('TabBar Sample'),
          ),
          body: const Column(
            children: <Widget>[
              Text('Scrollable-TabAlignment.start'),
              TabBar(
                isScrollable: true,
                tabAlignment: TabAlignment.start,
                tabs: <Widget>[
                  Tab(
                    icon: Icon(Icons.cloud_outlined),
                  ),
                  Tab(
                    icon: Icon(Icons.beach_access_sharp),
                  ),
                  Tab(
                    icon: Icon(Icons.brightness_5_sharp),
                  ),
                ],
              ),
              Text('Scrollable-TabAlignment.startOffset'),
              TabBar(
                isScrollable: true,
                tabAlignment: TabAlignment.startOffset,
                tabs: <Widget>[
                  Tab(
                    icon: Icon(Icons.cloud_outlined),
                  ),
                  Tab(
                    icon: Icon(Icons.beach_access_sharp),
                  ),
                  Tab(
                    icon: Icon(Icons.brightness_5_sharp),
                  ),
                ],
              ),
              Text('Scrollable-TabAlignment.center'),
              TabBar(
                isScrollable: true,
                tabAlignment: TabAlignment.center,
                tabs: <Widget>[
                  Tab(
                    icon: Icon(Icons.cloud_outlined),
                  ),
                  Tab(
                    icon: Icon(Icons.beach_access_sharp),
                  ),
                  Tab(
                    icon: Icon(Icons.brightness_5_sharp),
                  ),
                ],
              ),
              Spacer(),
              Text('Non-scrollable-TabAlignment.fill'),
              TabBar(
                tabAlignment: TabAlignment.fill,
                tabs: <Widget>[
                  Tab(
                    icon: Icon(Icons.cloud_outlined),
                  ),
                  Tab(
                    icon: Icon(Icons.beach_access_sharp),
                  ),
                  Tab(
                    icon: Icon(Icons.brightness_5_sharp),
                  ),
                ],
              ),
              Text('Non-scrollable-TabAlignment.center'),
              TabBar(
                tabAlignment: TabAlignment.center,
                tabs: <Widget>[
                  Tab(
                    icon: Icon(Icons.cloud_outlined),
                  ),
                  Tab(
                    icon: Icon(Icons.beach_access_sharp),
                  ),
                  Tab(
                    icon: Icon(Icons.brightness_5_sharp),
                  ),
                ],
              ),
              Spacer(),
            ],
          ),
          floatingActionButton: FloatingActionButton.extended(
            onPressed: () {
              setState(() {
                rtl = !rtl;
              });
            },
            label: const Text('Switch Direction'),
            icon: const Icon(Icons.swap_horiz),
          ),
        ),
      ),
    );
  }
}
``` 
	
</details>

![Screenshot 2023-06-06 at 18 06 12](https://github.com/flutter/flutter/assets/48603081/5ee5386d-cc64-4025-a020-ed2222cb6031)



## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

Piinks reviewed (commented):
Looks like there are a bunch o failing checks currently, is that because it is waiting on https://github.com/flutter/flutter/pull/125883?

TahaTesser commented:
> Looks like there are a bunch o failing checks currently, is that because it is waiting on #125883?

Yep, this is not expected to be clear until https://github.com/flutter/flutter/pull/125883 is merged. I've split this and tested it locally.

TahaTesser commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -3301,8 +3300,9 @@ void main() {
               children: <TestSemantics>[
                 TestSemantics(
                     id: 3,
-                    rect: TestSemantics.fullScreen,
+                    rect: const Rect.fromLTRB(0.0, 0.0, 232.0, 48.0),
                     flags: <SemanticsFlag>[SemanticsFlag.hasImplicitScrolling],
+                    transform: Matrix4.translationValues(0.0, 276.0, 0.0),
```
This semantic test is expected to be updated because this PR wraps tabs and tab ndicator in a `Stack` widget (divider), which updates the layout to be within the `Stack`.


### Before
| Indicator is aligned with the tab | Inspector view |
| --------------- | --------------- |
| <img src="https://user-images.githubusercontent.com/48603081/236506506-4a618e7a-287b-4a8f-89b3-57244f181241.png" /> | <img src="https://user-images.githubusercontent.com/48603081/236506563-10706a43-acb7-4fdb-b0c0-44fcedf41d31.png" /> |



| Indicator is aligned with the tab after introducing the `Stack` Widget |  Inspector view |
| --------------- | --------------- |
| <img src="https://user-images.githubusercontent.com/48603081/236506647-3bd4b6e9-c388-4a48-98f9-c3715093624e.png" /> | <img src="https://user-images.githubusercontent.com/48603081/236506657-0cb99a83-0709-4382-b1d1-2689c9a58e8f.png" /> |

TahaTesser commented on dev/tools/gen_defaults/lib/tabs_template.dart:
```diff
@@ -71,7 +74,7 @@ class _${blockName}PrimaryDefaultsM3 extends TabBarTheme {
   InteractiveInkFeatureFactory? get splashFactory => Theme.of(context).splashFactory;
 
   @override
-  TabAlignment? get tabAlignment => isScrollable ? TabAlignment.start : TabAlignment.fill;
+  TabAlignment? get tabAlignment => isScrollable ? TabAlignment.startOffset : TabAlignment.fill;
```
Use the newly added tab alignment to match the material specs for the scrollable tab bar. 


![Screenshot 2023-05-05 at 18 54 00](https://user-images.githubusercontent.com/48603081/236507223-9f4ef4f2-1d8f-48ae-9676-b93cdc354808.png)

Piinks replied:
I think we should run some extra testing on this to ensure it isn't reverted later. 👍

Piinks commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -1634,6 +1637,28 @@ class _TabBarState extends State<TabBar> {
       );
     }
 
+    final AlignmentGeometry effectiveAlignment = effectiveTabAlignment == TabAlignment.center
```
This should be a switch so the analyzer catches this case if TabAlignment ever changes.
Thanks to Dart 3 though, it can be super fun like:

```dart
final AlignmentGeometry effectiveAlignment = switch (effectiveTabAlignment) {
      TabAlignment.center => Alignment.center,
      TabAlignment.start || TabAlignment.startOffset || TabAlignment.fill => AlignmentDirectional.centerStart,
    };
```

TahaTesser replied:
This looks cool

Piinks replied:
Right? I like it.

Piinks reviewed (approved):
LGTM with the switch statement nit outstanding. I will fire off some extra internal tests, hold off on merging right away. :)

Piinks commented:
I am working on migrating customers here. For now, can you hold off on pushing any commits? The Google tests throw out the old result when a new commit is pushed. 😋

Piinks commented:
In the meantime, I think a short migration guide in flutter/website is warranted here. When folks upgrade to this and the tab bar moves, it may not be easy to figure out why or how to restore it.

TahaTesser commented:
> > In the meantime, I think a short migration guide in flutter/website is warranted here. When folks upgrade to this and the tab bar moves, it may not be easy to figure out why or how to restore it.
> 
> Great idea, I'll work on this.

Filed a PR for migration guide https://github.com/flutter/website/pull/8700

Piinks commented:
Thanks again for your patience here! I am still working through the customer tests. I found one case where it looks like the indicator and divider are not properly aligned when updated with this PR. I am still looking into why, but wanted to chare in case anything jumped out at you. I'll continue investigating.

I have omitted the labels since it is an internal customer. Ignore the horizontal alignment, the images here are cropped. It is the little bit of vertical spacing that is the issue:

This is the expected behavior:

![Screenshot 2023-05-23 at 12 27 39 PM](https://github.com/flutter/flutter/assets/16964204/0eaf6890-9530-495f-8837-c2d4f541a37f)

This PR causes this:

![Screenshot 2023-05-23 at 12 27 44 PM](https://github.com/flutter/flutter/assets/16964204/584dd5ff-d9c7-4d68-9065-0252ce976edf)

Piinks commented:
If you are up for it, splitting this up further may help. There are hundreds of image tests and some of them are affected by the divider height change, some by the tab alignment, it makes it more challenging to sort through them all. Just came across another odd case:

Expected:

![Screenshot 2023-05-23 at 12 32 15 PM](https://github.com/flutter/flutter/assets/16964204/91ca5b8b-a6bf-4235-bdb5-b148c489ecef)

With this PR:

![Screenshot 2023-05-23 at 12 32 20 PM](https://github.com/flutter/flutter/assets/16964204/1f141aa0-c620-4dbb-8d72-c8245b7c5796)

TahaTesser commented:
@Piinks 
Thank you for the details, I really appreciate it. The change is small but tricky to verify if it is fixing the problems that you shared. 

Is it possible for you to share the tab bar layout code from [this test](https://github.com/flutter/flutter/pull/125974#issuecomment-1559873470) above so I can test my fix in the same conditions? 
 It will be tremendously helpful for me to test rather than guess what might be going wrong there.

Piinks commented:
Ok, so for that specific test, it looks like the customer has a custom AppBar and a custom TabBar inside of that.
But, it looks like this might be what you are looking for, the custom TabBar is this:

```dart
Stack(
    alignment: Alignment.bottomCenter,
    children: [
      Divider(height: 0),
      TabBar(
        controller: tabController,
        labelPadding: EdgeInsets.zero,
        tabs: tabs,
        onTap: (tappedTabIndex) {
          //...
        },
      ),
    ],
  );
```

Does this help? If not I can dig further, just let me know! :)

Piinks commented:
Hmm, still trying to get a reproduction here. The fact that the customer is using a custom app bar and a custom tab bar is making it a bit difficult. Just wanted to let you know I was still looking into this.

TahaTesser commented:
> Hmm, still trying to get a reproduction here. The fact that the customer is using a custom app bar and a custom tab bar is making it a bit difficult. Just wanted to let you know I was still looking into this.

Appreciate it!

TahaTesser commented:
Friendly ping

Piinks commented:
I have tried a bunch of different ways to try to repro the customer case, but have not been successful. I am waiting to hear back from them to see if we can remove the strange extra Divider, because it seems redundant on top of the divider that is already rendered by the TabBar.

@TahaTesser you mentioned you had an alternative solution? We could also try that and see if it resolves the issue.

@HansMuller may also have a recommendation here since we cannot reproduce the customer failure.

TahaTesser committed:
Fix Material 3 Scrollable `TabBar` and `dividerHeight`

TahaTesser commented:
@Piinks 
First of all, I really appreciate your time on this. 

Based on the screenshots you shared, my best guess for the problem is constrained, so I went to the safest route and updated the current divider to be drawn for the whole width. This way layout is not modified, it's very close to the current behavior on the master. (current tab indicator is also drawn like this, it's not constrained as well)

Looking forward to knowing if this safer approach is breaking customer tests or not.

I also added a complete code sample in the description for anyone running into this PR.

Piinks commented:
Thank you! I'll run this through and see if it still occurs. Much appreciated as always!

Piinks commented:
Update - all customers should be migrated now. I am rerunning the tests and will need to go through all the images again just to make sure there are no more unexpected ones.

Piinks commented:
This is safe to land now, but the google testing shard is stuck. I filed a bug internally. Rebasing to see if that clears it up.

Piinks reviewed (commented):
Ah no I rebased and now there are some analysis issues, can you take a look?

TahaTesser commented:
> Ah no I rebased and now there are some analysis issues, can you take a look?

Taking a look.

TahaTesser committed:
Update template to use the new `getToken` method

TahaTesser committed:
Fix secondary divider height token un-used

Piinks commented:
Hmm looks like we have some new test failures here

TahaTesser commented:
> Hmm looks like we have some new test failures here

That's weird. I can reproduce this locally too. Pretty sure we had most of it green before, didn't we?
I'll take a closer look in the morning.

TahaTesser committed:
Explicitly enable `useMaterial3` flag in the M3 tests

TahaTesser committed:
Use `theme.useMaterial3` instead of hard-coded value

TahaTesser commented:
@Piinks 
Mystery solved! 

https://github.com/flutter/flutter/pull/128725 recently updated the `useMaterial3` flag in the boilerplate code to be nullable in the material library for M2. 
I just had to explicitly enable `theme.useMaterial3`  for the M3 tests in this PR.

TahaTesser commented:
`Linux flutter_plugins ` is failing on most PRs right now.

```console

[31mThe following packages had errors:[0m
[31m  file_selector/file_selector[0m
[31m  file_selector_linux[0m
[31m  file_selector_macos[0m
[31m  file_selector_windows[0m
[31mSee above for full details.[0m
```

HansMuller reviewed (approved):
LGTM

Piinks commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -1269,8 +1284,10 @@ class _TabBarState extends State<TabBar> {
       labelPaddings: _labelPaddings,
+      dividerColor: theme.useMaterial3 ? widget.dividerColor ?? tabBarTheme.dividerColor ?? _defaults.dividerColor : null,
+      dividerHeight: theme.useMaterial3 ? widget.dividerHeight ?? tabBarTheme.dividerHeight ?? _defaults.dividerHeight : null,
+      width: MediaQuery.sizeOf(context).width,
```
@TahaTesser do you recall why this changed? I do not remember since this PR has had so much back and forth, I am looking though..

TahaTesser replied:
You mean the `width`? This what divider painters uses to draw the divider beyond the tabs without changing the layout.

TahaTesser replied:
Previously divider was only drawn with in the tabs, this caused this issue. https://github.com/flutter/flutter/issues/117722

Piinks replied:
Right right. Thank you for the reference! I could not recall. We're reviewing all of the screenshots one last time and there was a weird one where the customer mocked up a phone outline in the test and the divider extended out beyond it. I think it is an unintended issue of the test, not this change. No worries. :)

Piinks replied:
Ok so sad news, I am going to revert this. Unfortunately this is not uncommon when there are over a thousand screenshots. 😢 

I misunderstood what an image was capturing. The width here actually causes an issue when the TabBar is contained within a ConstrainedBox, the divider extends beyond it.

The customer test was not showing a mocked up phone, but rather a floating box with the content it was floating over omitted from the test case.

I've put together a repro here:

![Screenshot 2023-06-22 at 5 28 26 PM](https://github.com/flutter/flutter/assets/16964204/fb8afc61-2759-4bcc-82ae-4b3b5d9006c5)


```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {

    return MaterialApp(
      theme: ThemeData(useMaterial3: true),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {

  const MyHomePage({ super.key });

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: DefaultTabController(
        length: 2,
        child: Center(
          child: ConstrainedBox(
            constraints: const BoxConstraints(maxWidth: 360),
            child: ColoredBox(
              color: Colors.grey[200]!,
              child: const TabBar.secondary(
                tabAlignment: TabAlignment.start,
                isScrollable: true,
                tabs: [
                  Tab(text: 'Test 1'),
                  Tab(text: 'Test 2'),
                ],
              ),
            )
          ),
        ),
      ),
    );
  }
}
```
